### PR TITLE
Refactor IR and its construction

### PR DESCRIPTION
**Pull request by yloiseau**
This is a refactoring to ease macro implementations.

- makes IR nodes mutable (in a limited way)
- makes a builder-like fluent interface to IR nodes
- better encapsulation by moving methods in the appropriate node
- adds a walk method to node to ease the visit (internal iteration)
- abstract dummy IR visitor to ease the creation of new specific
  visitors
- refactoring of IR visitors to use the abstract one
- introduction of new IR nodes (e.g. augmentations,
  ForEachLoop) to move some syntactic sugar from AST to IR transformation
  into a separate IR visitor (this will allow macros to build such node
  more easily)
- adds Builders to create a DSL to ease the creation of the IR
- refactoring of AST to IR visitor to use the new DSL and fluent API

Also clean some code according to PMD and checkstyle

### CODE REVIEW

**Comment by jponge** [comment]
This is big one Yannick :-)

I’ll have a look soon.

Thanks!

**Comment by jponge** [comment]
I'm doing a quick pass now.

File: src/main/java/gololang/Observable.java [hunk 1]
```diff
@@ -109,15 +109,15 @@ public String toString() {
         '}';
   }
 
-  public static interface Function {
+  public interface Function {
```
Comment (jponge) [inline src/main/java/gololang/Observable.java]: Why did you remove `static` from the interface definitions?

File: src/main/java/gololang/Predefined.java [hunk 2]
```diff
@@ -598,7 +603,7 @@ public static boolean isArray(Object object) {
    * @return the Character value.
    * @throws IllegalArgumentException if {@code obj} is not a number or a String.
    */
-  public static Object charValue(Object obj) throws IllegalArgumentException {
+  public static Object charValue(Object obj) {
```
Comment (jponge) [inline src/main/java/gololang/Predefined.java]: Why removing the `throws` declarations?

File: src/main/java/org/eclipse/golo/compiler/JavaBytecodeGenerationGoloIrVisitor.java [hunk 3]
```diff
@@ -79,56 +52,89 @@
...
+    throw invalidElement(match);
+  }
+
+  @Override
+  public void visitCaseStatement(CaseStatement caseStatement) {
+    throw invalidElement(caseStatement);
+  }
+
+  @Override
+  public void visitWhenClause(WhenClause<?> whenClause) {
```
Comment (jponge) [inline src/main/java/org/eclipse/golo/compiler/JavaBytecodeGenerationGoloIrVisitor.java]: Why is there a generic?

File: src/main/java/org/eclipse/golo/compiler/JavaBytecodeGenerationGoloIrVisitor.java [hunk 4]
```diff
@@ -220,34 +228,47 @@ private void writeAugmentationApplicationsMetaData(Map<String, Set<String>> appl
...
+        imports.stream()
+        .map(ModuleImport::getPackageAndClass)
+        .map(PackageAndClass::toString)
+        .toArray(String[]::new));
   }
 
-  private void writeAugmentsMetaData(Set<String> augmentations) {
-    String[] augmentArray = augmentations.toArray(new String[augmentations.size()]);
-    writeMetaData("augmentations", augmentArray);
+  private void writeAugmentsMetaData() {
```
Comment (jponge) [inline src/main/java/org/eclipse/golo/compiler/JavaBytecodeGenerationGoloIrVisitor.java]: :+1: :smiley:

File: src/main/java/org/eclipse/golo/compiler/ir/ConstantStatement.java [hunk 5]
```diff
@@ -11,9 +11,9 @@
 
 public class ConstantStatement extends ExpressionStatement {
 
-  private final Object value;
+  private Object value;
 
-  public ConstantStatement(Object value) {
+  ConstantStatement(Object value) {
```
Comment (jponge) [inline src/main/java/org/eclipse/golo/compiler/ir/ConstantStatement.java]: Why not `public` anymore?

File: src/main/java/org/eclipse/golo/compiler/ir/WhenClause.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.golo.compiler.ir;
+
+public final class WhenClause<T extends GoloElement> extends GoloElement {
```
Comment (jponge) [inline src/main/java/org/eclipse/golo/compiler/ir/WhenClause.java]: Ok, got it wrt previous comment.

File: src/main/java/gololang/Predefined.java (hunk 2 above)
Comment (yloiseau) [inline src/main/java/gololang/Predefined.java]: Not mandatory since it's an unchecked exception (pmd was complaining). But it can indeed be useful as a documentation purpose

File: src/main/java/org/eclipse/golo/compiler/ir/ConstantStatement.java (hunk 5 above)
Comment (yloiseau) [inline src/main/java/org/eclipse/golo/compiler/ir/ConstantStatement.java]: To "force" the use of the corresponding `Builders` method

File: src/main/java/gololang/Observable.java (hunk 1 above)
Comment (yloiseau) [inline src/main/java/gololang/Observable.java]: It's redundant (a nested interface is automatically "static")

File: src/main/java/gololang/Predefined.java (hunk 2 above)
Comment (jponge) [inline src/main/java/gololang/Predefined.java]: > Not mandatory since it's an unchecked exception (pmd was complaining). But it can indeed be useful as a documentation purpose

In that case PMD is wrong :smile:

(note that I don’t like PMD)

File: src/main/java/org/eclipse/golo/compiler/ir/ConstantStatement.java (hunk 5 above)
Comment (jponge) [inline src/main/java/org/eclipse/golo/compiler/ir/ConstantStatement.java]: Ok

File: src/main/java/gololang/Observable.java (hunk 1 above)
Comment (jponge) [inline src/main/java/gololang/Observable.java]: Oh yes, right!

File: src/main/java/gololang/Predefined.java (hunk 2 above)
Comment (yloiseau) [inline src/main/java/gololang/Predefined.java]: Just curious, why do you say it's wrong? Because of the documenting value of being explicit?
If so I agree with that point, although is in the javadoc :smile:
Anyway, I'll put it back. I've removed such explicit `throws` of `RuntimeExceptions` in several places, should I rollback everywhere?

Comment (jponge) [inline src/main/java/gololang/Predefined.java]: Yes, I value the explicitness for unchecked exceptions that may still likely be thrown. Javadoc is one thing, yet `throws` does, too.

The exception is probably `RuntimeException` which is very generic, but all of that is me :smile:

**Comment by jponge** [comment]
@yloiseau Looks good for me
